### PR TITLE
ASoC: SOF: Add support for IPC4 via the SOF client infrastructure

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

Prepare the SOF client core code to process IPC4 messages for both tx and rx (notifications).
Add APIs  for the clients to check the IPC version when they do runtime decision on how to handle the messages.
Add support for IPC4 in the message injector.

### CODE REVIEW

File: sound/soc/sof/sof-client-ipc-msg-injector.c [hunk 1]
```diff
@@ -136,8 +137,9 @@ static int sof_msg_inject_probe(struct auxiliary_device *auxdev,
 	if (!priv)
 		return -ENOMEM;
 
-	priv->tx_buffer = devm_kmalloc(dev, SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
-	priv->rx_buffer = devm_kzalloc(dev, SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
+	priv->max_msg_size = sof_client_get_ipc_max_payload_size(cdev);
```
Comment (plbossart) [inline sound/soc/sof/sof-client-ipc-msg-injector.c:262]: can there be an error here?

**Review by plbossart — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

File: sound/soc/sof/sof-client-ipc-msg-injector.c (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/sof-client-ipc-msg-injector.c:262]: No. If the max_payload_size would be invalid then the SOF stack would have failed already so we would not have this (or other) clients loaded.

**Comment by plbossart** [comment]
Up2 configuration is borked in https://sof-ci.01.org/linuxpr/PR3470/build7251/devicetest/, merging anyways.

**Comment by ujfalusi** [comment]
> Up2 configuration is borked in https://sof-ci.01.org/linuxpr/PR3470/build7251/devicetest/, merging anyways.

Hrm, hw:0,0 PCM is not created on up2-hda. I don't think it can be caused by this PR, but then what causes it?

**Comment by plbossart** [comment]
> > Up2 configuration is borked in https://sof-ci.01.org/linuxpr/PR3470/build7251/devicetest/, merging anyways.
> 
> Hrm, hw:0,0 PCM is not created on up2-hda. I don't think it can be caused by this PR, but then what causes it?

@ujfalusi it uses a generic HDaudio topology instead of idisp/HDMI only!

**Comment by plbossart** [comment]
@ujfalusi look at the dmesg log:

````
[   10.108219] kernel: sof-audio-pci-intel-apl 0000:00:0e.0: loading topology:intel/sof-tplg/sof-hda-generic-idisp.tplg
````

but CI uses sof-hda-generic.tplg. 

````
TPLG=/lib/firmware/intel/sof-tplg/sof-hda-generic.tplg MODEL=APL_UP2_HDA ~/sof-test/test-case/verify-kernel-boot-log.sh
````

Well d'uh, if the tests use the wrong file then of course they will flag errors with mismatches between PCM devices created and expected.

@keqiaozhang @XiaoyunWu6666 @mengdonglin @fredoh9 @greg-intel FYI

**Comment by keqiaozhang** [comment]
> Well d'uh, if the tests use the wrong file then of course they will flag errors with mismatches between PCM devices created and expected.
> 
> @keqiaozhang @XiaoyunWu6666 @mengdonglin @fredoh9 @greg-intel FYI

I think @XiaoyunWu6666 has updated tplg name in our CI model configuration this morning.
 
image

**Pull request merged**